File: app/src/main/java/com/mytest/widget/layoutmanagerjava/OverLayCardItemCallback.java
package com.mytest.widget.layoutmanagerjava;

import android.graphics.Canvas;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

/**
 * 介绍：人人影视效果的Callback
 * 作者：zhangxutong
 * 邮箱：mcxtzhang@163.com
 * 主页：http://blog.csdn.net/zxt0601
 * 时间： 16/12/18.
 */

public class OverLayCardItemCallback extends ItemTouchHelper.SimpleCallback {

    protected RecyclerView mRv;
    protected List mDatas;
    protected RecyclerView.Adapter mAdapter;

    public OverLayCardItemCallback(RecyclerView rv, RecyclerView.Adapter adapter, List datas) {
//        this(0,
//                ItemTouchHelper.DOWN | ItemTouchHelper.UP | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT,
//                rv, adapter, datas);

        this(0,
                ItemTouchHelper.LEFT,
                rv, adapter, datas);
    }

    public OverLayCardItemCallback(int dragDirs, int swipeDirs
            , RecyclerView rv, RecyclerView.Adapter adapter, List datas) {
        super(dragDirs, swipeDirs);
        mRv = rv;
        mAdapter = adapter;
        mDatas = datas;
    }

    //水平方向是否可以被回收掉的阈值
    public float getThreshold(RecyclerView.ViewHolder viewHolder) {
        //2016 12 26 考虑 探探垂直上下方向滑动，不删除卡片，这里参照源码写死0.5f
//        return mRv.getWidth() * /*getSwipeThreshold(viewHolder)*/ 0.5f;
        return mRv.getWidth() * getSwipeThreshold(viewHolder);
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        return false;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        //Log.e("swipecard", "onSwiped() called with: viewHolder = [" + viewHolder + "], direction = [" + direction + "]");
        //rollBack(viewHolder);
        //★实现循环的要点
        Object remove = mDatas.remove(viewHolder.getLayoutPosition());
        mDatas.add(0, remove);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void onChildDraw(Canvas canvas, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
        super.onChildDraw(canvas, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        //先根据滑动的dxdy 算出现在动画的比例系数fraction
        double swipValue = Math.sqrt(dX * dX + dY * dY);
        double fraction = swipValue / getThreshold(viewHolder);
        //边界修正 最大为1
        if (fraction > 1) {
            fraction = 1;
        }
        //对每个ChildView进行缩放 位移
        int childCount = recyclerView.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = recyclerView.getChildAt(i);
            //第几层,举例子，count =7， 最后一个TopView（6）是第0层，
            int level = childCount - i - 1;
            if (level > 0) {
                child.setScaleY((float) (1 - CardConfig.SCALE_GAP * level + fraction * CardConfig.SCALE_GAP));

                if (level < CardConfig.MAX_SHOW_COUNT - 1) {
                    child.setScaleX((float) (1 - CardConfig.SCALE_GAP * level + fraction * CardConfig.SCALE_GAP));
                    child.setTranslationX((float) (CardConfig.TRANS_Y_GAP * level - fraction * CardConfig.TRANS_Y_GAP));
                } else {
                    //child.setTranslationY((float) (mTranslationYGap * (level - 1) - fraction * mTranslationYGap));
                }
            }
        }

        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE){
            float value = 1 - Math.abs(dX) / viewHolder.itemView.getWidth();
            viewHolder.itemView.setAlpha(value);
            viewHolder.itemView.setScaleY(value);
        }
    }

    @Override
    public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);

        viewHolder.itemView.setAlpha(1);
        viewHolder.itemView.setScaleY(1);
    }
}

File: app/src/main/java/com/mytest/widget/layoutmanager/OverLayCardItemCallback.kt
package com.mytest.widget.layoutmanager

import android.content.Context
import android.graphics.Canvas
import android.util.Log
import android.util.TypedValue
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import kotlin.math.abs
import kotlin.math.sqrt

class OverLayCardItemCallback @JvmOverloads constructor(
        dragDirs: Int,
        swipeDirs: Int,
        private val recyclerView: RecyclerView,
        private val adapter: RecyclerView.Adapter<*>,
        private var datas: MutableList<Any>)
    : ItemTouchHelper.SimpleCallback(dragDirs, swipeDirs) {

    companion object{
        const val TAG = "OverLayCardItemCallback"
    }

    //设置水平方向是否可以被回收掉的阀值
    private fun getThreshold(viewHolder: RecyclerView.ViewHolder): Float{
        //写了一个固定的值，如果滑动的宽度是整体宽度的三分之一，就回收
        Log.d(TAG, "recyclerView width: ${recyclerView.width}")
        return recyclerView.width * 0.33f
    }

    override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        Log.d(TAG, "onMove")
        //viewHolder当前被拖动的item, target当前被拖动的item下方的item
        return false
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        //删除当前拖动的item,把删除的数组放入底部，实现无限循环
        val remove = datas.removeAt(viewHolder.layoutPosition)
        datas.remove(remove)
        datas.add(remove)
        adapter.notifyDataSetChanged()
    }

    override fun onChildDraw(canvas: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
        super.onChildDraw(canvas, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)

        //当前移动的item进行缩放与透明化
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            val value = 1 - abs(dX) / viewHolder.itemView.width
            viewHolder.itemView.alpha = value
            viewHolder.itemView.scaleY = value
        }

        val childCount = recyclerView.childCount
        Log.d("onChildDraw", "$childCount")

        //根据滑动的dx与dy，算出现在动画的比例系数
//        val swipedValue = sqrt(dX * dX + dY * dY)
//        var fraction = swipedValue / getThreshold(viewHolder)
//        Log.d(TAG, "滑动的系数是：${fraction}")
//        //边界修正
//        if (fraction > 1){
//            fraction = 1f
//        }
        //对每个childView进行缩放，位移
//        val childCount = recyclerView.childCount
//        for (index in 0 until childCount){
//            val child = recyclerView.getChildAt(index)
//            val level = childCount - index - 1
//            if (level > 0){
//                child.scaleY = 1 - Constants.SCALE_GAP * level + fraction * Constants.SCALE_GAP
//
//                if (level < childCount - 1){
//                    child.scaleX = 1 - Constants.SCALE_GAP * level + fraction * Constants.SCALE_GAP
//                    child.translationX = getDisplayMetrics(child.context) * level - fraction * getDisplayMetrics(child.context)
//                }
//            }
//        }


    }

    private fun getDisplayMetrics(context: Context): Int{
        Log.d("getDisplayMetrics", "Callback display:${TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, Constants.TRANS_X_GAP, context.resources.displayMetrics).toInt()}")
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, Constants.TRANS_X_GAP, context.resources.displayMetrics).toInt()
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: ViewHolder) {
        super.clearView(recyclerView, viewHolder)
        viewHolder.itemView.alpha = 1f
        viewHolder.itemView.scaleY = 1f
    }
}